neon_messagebus_mq_connector/controller.py

```diff
@@ -134,89 +149,144 @@ def handle_neon_profile_update(self, message: Message):
+
+        LOG.info('Initiating template validation')
+        for message_template in message_templates:
             try:
-                for message_template in message_templates:
-                    dict_data = message_template(**dict_data).dict()
+                msg_data = message_template(**msg_data).dict()
             except (ValueError, ValidationError) as err:
-                return err, dict_data
-            return None, dict_data
-
-        # TODO: This is really `templates`, not `skills`
-        request_skills = dict_data["context"].get("request_skills",
-                                                  ["default"])
-        if len(request_skills) == 0:
-            request_skills = ["default"]
-        try:
-            message_templates = [templates[request_type]
-                                 for request_type in request_skills]
-        except KeyError:
-            return None, dict_data
-        check_error, dict_data = check_keys_presence(dict_data,
-                                                     message_templates)
-        return check_error, dict_data
+                LOG.error(f'Failed to validate {msg_data} with template = {message_template.__name__}, exception={err}')
+                return str(err), msg_data
+        LOG.info('Template validation completed successfully')
+        return '', msg_data
+
+    @classmethod
+    def validate_request(cls, msg_data: dict):
+        """
+            Fetches the relevant template models and validates provided message data iteratively through them
+
+            :param msg_data: message data for validation
+
+            :return: validation details(None if validation passed),
+                     input data with proper data types and filled default fields
+        """
+
+        requested_templates = msg_data.get("context", {}).get("request_skills") or ["default"]
+        message_templates = []
+
+        for requested_template in requested_templates:
+            matching_template_model = templates.get(requested_template)
+            if not matching_template_model:
+                LOG.warning(f'Template under keyword "{requested_template}" does not exist')
+            else:
+                message_templates.append(matching_template_model)
+
+        detected_error, msg_data = cls.__validate_message_templates(msg_data=msg_data,
+                                                                    message_templates=message_templates)
+        return detected_error, msg_data
+
+    def push_awaiting_message(self, message: Message):
+        """ Pushes awaiting message so its relevant data could be fetched once received response """
+        message_id = message.context.get('mq', {}).get('message_id')
+        if not message_id:
+            LOG.warning('Awaiting message not pushed - message_id is None')
+        else:
+            if message.msg_type == 'neon.get_stt':
+                self.awaiting_messages['stt'][message_id] = {'lang': message.data['lang'],
```

Comment:
As noted below, consider adding this data to message context and just tracking a list of `message_id` internally